src/cargo/util/config/mod.rs

```diff
@@ -2180,6 +2180,57 @@ pub struct CargoBuildConfig {
     pub out_dir: Option<ConfigRelativePath>,
 }
 
+/// Configuration for `build.target`.
+///
+/// Accepts in the following forms:
+///
+/// ```toml
+/// target = "a"
+/// target = ["a"]
+/// target = ["a", "b"]
+/// ```
+#[derive(Debug, Deserialize)]
+#[serde(transparent)]
+pub struct BuildTargetConfig {
+    inner: Value<BuildTargetConfigInner>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+enum BuildTargetConfigInner {
+    One(String),
+    Many(Vec<String>),
+}
+
+impl BuildTargetConfig {
+    /// Gets values of `build.target` as a list of [`BuildTargetConfigValue`].
+    pub fn values(&self, config: &Config) -> Vec<BuildTargetConfigValue<'_>> {
```

Comment:
Can this perhaps be simplified by returning a `Vec<String>` instead of a separate type?  I don't think it is particularly necessary to return a separate enum type which then immediately gets unpacked back to a string.

Also, that could help reduce the code duplication in `from_requested_targets`.  Ideally that dancing around with BTreeSet and creating CompileTarget could be unified.